crates/revm/src/state_change.rs

```diff
@@ -51,11 +56,127 @@ pub fn post_block_balance_increments(
     balance_increments
 }
 
-/// Applies the pre-block call to the EIP-4788 beacon block root contract, using the given block,
+/// todo: temporary move over of constants from revm until we've migrated to the latest version
+pub const HISTORY_SERVE_WINDOW: usize = 8192;
+
+/// Applies the pre-block state change outlined in [EIP-2935] to store historical blockhashes in a
+/// system contract.
+///
+/// If Prague is not activated, or the block is the genesis block, then this is a no-op, and no
+/// state changes are made.
+///
+/// If the provided block is the fork activation block, this will generate multiple state changes,
+/// as it inserts multiple historical blocks, as outlined in the EIP.
+///
+/// If the provided block is after Prague has been activated, this will only insert a single block
+/// hash.
+///
+/// [EIP-2935]: https://eips.ethereum.org/EIPS/eip-2935
+#[inline]
+pub fn apply_blockhashes_update<DB: Database + DatabaseCommit>(
+    chain_spec: &ChainSpec,
+    block_timestamp: u64,
+    block_number: u64,
+    db: &mut DB,
+) -> Result<(), BlockExecutionError>
+where
+    DB::Error: std::fmt::Display,
+{
+    // If Prague is not activated or this is the genesis block, no hashes are added.
+    if !chain_spec.is_prague_active_at_timestamp(block_timestamp) || block_number == 0 {
+        return Ok(())
+    }
+    assert!(block_number > 0);
+
+    // We load the `HISTORY_STORAGE_ADDRESS` account because REVM expects this to be loaded in order
+    // to access any storage, which we will do below.
+    // If the account does not exist, we create it with the EIP-2935 bytecode and a nonce of 1, so
+    // it does not get deleted.
```

Comment:
```suggestion
    // Account is expected to exist either in genesis (for tests) or redeployed on the mainnet or testnets.
    // If the account for any reason does not exist, we create it with the EIP-2935 bytecode and a nonce of 1, so
    // it does not get deleted.
```

First part is not true anymore as from last changes to the EIP